pkg/k8sutil/status.go

```diff
@@ -254,3 +256,62 @@ func UpdateRollingUpgradeState(c client.Client, cluster *v1beta1.KafkaCluster, t
 	logger.Info("Rolling upgrade status updated", "status", timeStamp)
 	return nil
 }
+
+func UpdateListenerStatuses(c client.Client, cluster *v1beta1.KafkaCluster, logger logr.Logger, extListenerStatuses map[string]v1beta1.ListenerStatus) error {
+	typeMeta := cluster.TypeMeta
+
+	intListenerStatuses := createInternalListenerStatuses(cluster)
+
+	cluster.Status.ListenerStatuses = v1beta1.ListenerStatuses{
+		InternalListeners: intListenerStatuses,
+		ExternalListeners: extListenerStatuses,
+	}
+
+	err := c.Status().Update(context.Background(), cluster)
+	if apierrors.IsNotFound(err) {
+		err = c.Update(context.Background(), cluster)
+	}
+	if err != nil {
+		if !apierrors.IsConflict(err) {
+			return errors.WrapIf(err, "could not update listener status")
+		}
+		err := c.Get(context.TODO(), types.NamespacedName{
+			Namespace: cluster.Namespace,
+			Name:      cluster.Name,
+		}, cluster)
+		if err != nil {
+			return errors.WrapIf(err, "could not get config for updating listener status")
+		}
+
+		cluster.Status.ListenerStatuses = v1beta1.ListenerStatuses{
+			InternalListeners: intListenerStatuses,
+			ExternalListeners: extListenerStatuses,
+		}
+
+		err = c.Status().Update(context.Background(), cluster)
+		if apierrors.IsNotFound(err) {
+			err = c.Update(context.Background(), cluster)
+		}
+		if err != nil {
+			return errors.WrapIf(err, "could not update listener statuses")
+		}
+	}
+	// update loses the typeMeta of the config that's used later when setting ownerrefs
+	cluster.TypeMeta = typeMeta
+	logger.Info("updated listener statuses")
+	return nil
+}
+
+func createInternalListenerStatuses(kafkaCluster *v1beta1.KafkaCluster) map[string]v1beta1.ListenerStatus {
+	intListenerStatuses := make(map[string]v1beta1.ListenerStatus)
```

Comment:
```suggestion
	intListenerStatuses := make(map[string]v1beta1.ListenerStatus,len(kafkaCluster.Spec.ListenersConfig.InternalListeners))
```

Even though this may pre-allocate a map that is larger by one IHMO it's still worth it.

Reply:
Agreed. That was my first thought as well.